src/main/drivers/serial_uart_stm32f7xx.c

```diff
@@ -361,140 +316,76 @@ void dmaIRQHandler(dmaChannelDescriptor_t* descriptor)
-    s->port.txBuffer = uart->txBuffer;
-    s->port.rxBufferSize = sizeof(uart->rxBuffer);
-    s->port.txBufferSize = sizeof(uart->txBuffer);
+    s->port.rxBuffer = uartdev->rxBuffer;
+    s->port.txBuffer = uartdev->txBuffer;
+    s->port.rxBufferSize = ARRAYLEN(uartdev->rxBuffer);
+    s->port.txBufferSize = ARRAYLEN(uartdev->txBuffer);
 
-    s->USARTx = uart->dev;
-    if (uart->rxDMAStream) {
-        s->rxDMAChannel = uart->DMAChannel;
-        s->rxDMAStream = uart->rxDMAStream;
+    const uartHardware_t *hardware = uartdev->hardware;
+
+    s->USARTx = hardware->reg;
+
+    if (hardware->rxDMAStream) {
+        s->rxDMAChannel = hardware->DMAChannel;
+        s->rxDMAStream = hardware->rxDMAStream;
     }
-    s->txDMAChannel = uart->DMAChannel;
-    s->txDMAStream = uart->txDMAStream;
+    s->txDMAChannel = hardware->DMAChannel;
+    s->txDMAStream = hardware->txDMAStream;
 
     s->txDMAPeripheralBaseAddr = (uint32_t)&s->USARTx->TDR;
     s->rxDMAPeripheralBaseAddr = (uint32_t)&s->USARTx->RDR;
 
-    s->Handle.Instance = uart->dev;
+    s->Handle.Instance = hardware->reg;
 
-    IO_t tx = IOGetByTag(uart->tx);
-    IO_t rx = IOGetByTag(uart->rx);
+    IO_t txIO = IOGetByTag(uartdev->tx);
+    IO_t rxIO = IOGetByTag(uartdev->rx);
 
-    if ((options & SERIAL_BIDIR) && tx) {
-        IOInit(tx, OWNER_SERIAL_TX, RESOURCE_INDEX(device));
-        IOConfigGPIOAF(tx, IOCFG_AF_PP, uart->af);
+    if ((options & SERIAL_BIDIR) && txIO) {
+        // XXX BIDIR_PP handling is missing
+        IOInit(txIO, OWNER_SERIAL_TX, RESOURCE_INDEX(device));
+        IOConfigGPIOAF(txIO, IOCFG_AF_PP, hardware->af);
     }
     else {
-        if ((mode & MODE_TX) && tx) {
-            IOInit(tx, OWNER_SERIAL_TX, RESOURCE_INDEX(device));
-            IOConfigGPIOAF(tx, IOCFG_AF_PP, uart->af);
+        if ((mode & MODE_TX) && txIO) {
+            IOInit(txIO, OWNER_SERIAL_TX, RESOURCE_INDEX(device));
+            IOConfigGPIOAF(txIO, IOCFG_AF_PP, hardware->af);
         }
 
-        if ((mode & MODE_RX) && rx) {
-            IOInit(rx, OWNER_SERIAL_RX, RESOURCE_INDEX(device));
-            IOConfigGPIOAF(rx, IOCFG_AF_PP, uart->af);
+        if ((mode & MODE_RX) && rxIO) {
+            IOInit(rxIO, OWNER_SERIAL_RX, RESOURCE_INDEX(device));
+            IOConfigGPIOAF(rxIO, IOCFG_AF_PP, hardware->af);
         }
     }
 
     // DMA TX Interrupt
-    dmaInit(uart->txIrq, OWNER_SERIAL_TX, RESOURCE_INDEX(device));
-    dmaSetHandler(uart->txIrq, dmaIRQHandler, uart->txPriority, (uint32_t)uart);
+    dmaInit(hardware->txIrq, OWNER_SERIAL_TX, RESOURCE_INDEX(device));
+    dmaSetHandler(hardware->txIrq, dmaIRQHandler, hardware->txPriority, (uint32_t)uartdev);
 
 
-    //HAL_NVIC_SetPriority(uart->txIrq, NVIC_PRIORITY_BASE(uart->txPriority), NVIC_PRIORITY_SUB(uart->txPriority));
-    //HAL_NVIC_EnableIRQ(uart->txIrq);
+    //HAL_NVIC_SetPriority(hardware->txIrq, NVIC_PRIORITY_BASE(hardware->txPriority), NVIC_PRIORITY_SUB(hardware->txPriority));
```

Comment:
I have some updated F7 uart code that's improves this.  When Keep these here for now please. My changes will be easier to merge.